frontend/javascripts/oxalis/model/sagas/min_cut_saga.js

```diff
@@ -0,0 +1,722 @@
+    if ((bitMask & (2 ** neighborIdx)) !== 0) {
+      neighbors.outgoing.push(NEIGHBOR_LOOKUP[neighborIdx]);
+    }
+    if ((bitMask & (2 ** (neighborIdx + NEIGHBOR_LOOKUP.length))) !== 0) {
+      neighbors.ingoing.push(NEIGHBOR_LOOKUP[neighborIdx]);
+    }
+  }
+
+  return neighbors;
+}
+const getNeighborsFromBitMask = _.memoize(_getNeighborsFromBitMask);
+
+// Functions to add/remove edges which mutate the bitmask.
+function addOutgoingEdge(edgeBuffer, idx, neighborIdx) {
+  edgeBuffer[idx] |= 2 ** neighborIdx;
+}
+function addIngoingEdge(edgeBuffer, idx, neighborIdx) {
+  edgeBuffer[idx] |= 2 ** (NEIGHBOR_LOOKUP.length + neighborIdx);
+}
+function removeIngoingEdge(edgeBuffer, idx, neighborIdx) {
+  edgeBuffer[idx] &= ~(2 ** (NEIGHBOR_LOOKUP.length + neighborIdx));
+}
+function removeOutgoingEdge(edgeBuffer, idx, neighborIdx) {
+  edgeBuffer[idx] &= ~(2 ** neighborIdx);
+}
+
+//
+// Actual min cut implementation.
+//
+// Note that a heuristic is used to
+// determine a appropriate mag to compute the min-cut in.
+// If the computation takes too long, the min-cut is aborted and a
+// poorer mag is tried.
+// Afterwards, it is tried to "refine" the min-cut by also calculating
+// the min-cut in better mags. As a result, the cut is initially drawn
+// "with broad/fast strokes" and the final details are solved in higher
+// mags.
+
+function* performMinCut(action: Action): Saga<void> {
+  if (action.type !== "PERFORM_MIN_CUT") {
+    throw new Error("Satisfy flow.");
+  }
+
+  const skeleton = yield* select(store => store.tracing.skeleton);
+  if (!skeleton) {
+    return;
+  }
+  const seedTree = skeleton.trees[action.treeId];
+
+  if (!seedTree) {
+    throw new Error("seedTree not found?");
+  }
+
+  const nodes = Array.from(seedTree.nodes.values());
+
+  if (nodes.length !== 2) {
+    // The min-cut operation should not be available in the context-menu when the tree
+    // does not have exactly two nodes.
+    return;
+  }
+
+  const boundingBoxes = skeleton.userBoundingBoxes.filter(bbox => bbox.isVisible);
+  let boundingBoxObj;
+  if (boundingBoxes.length === 0) {
+    console.log("No visible bounding box defined for min-cut. Creating one...");
+    const newBBox = {
+      min: V3.floor(V3.sub(V3.min(nodes[0].position, nodes[1].position), DEFAULT_PADDING)),
+      max: V3.floor(
+        V3.add(
+          V3.add(V3.max(nodes[0].position, nodes[1].position), DEFAULT_PADDING),
+          // Add [1, 1, 1], since BoundingBox.max is exclusive
+          [1, 1, 1],
+        ),
+      ),
+    };
+
+    yield* put(
+      addUserBoundingBoxAction({
+        boundingBox: newBBox,
+        name: "Bounding box used for splitting cell",
+        color: Utils.getRandomColor(),
+        isVisible: true,
+      }),
+    );
+
+    boundingBoxObj = newBBox;
+  } else if (boundingBoxes.length === 1) {
+    boundingBoxObj = boundingBoxes[0].boundingBox;
+  } else {
+    console.log(
```

Comment:
> My gut tells me that this bounding box business is a little finicky and opaque. I'm not sure whether users will understand the effect of having bounding boxes visible, since it doesn't seem to be communicated. I can also imagine that it is somewhat annoying having to disable all (but one) bounding boxes every time a min-cut should be performed OR users forgetting they have an unrelated bounding box visible.

You are absolutely right. This was bugging me, too. I now changed the code so that one can choose which bounding box should be used from the context menu. Only the bounding boxes which contain the seed nodes are shown in the list. Additionally, one can select "Create new bounding box". I'm quite happy with the UX now :)
The sad news is that this change required to use antd submenus within the context menu which turned out to be very painful. I had to dig quite deep and tried several approaches to get this working. The final solution can be seen in the following commits. The link points to a comment where I explained the rationale behind this. I suggest to review these commits separately. Also note that I had to move quite a lot of code within the context menu module. Most of it stayed as is, but I added a new `<Dropdown />` component and separated it from the actual container. The upside is that I could remove quite a lot of custom CSS classes which we passed to antd menus :tada: 

https://github.com/scalableminds/webknossos/pull/5885/files/600dc941fe5e5a251cba60112b23fdb3fc846f78..58562dd90b1c73bc35b35cafe54215ee04fdf56d#diff-989748590830a0e73eef9fd9e98df428dab5559c0a88e4a656e9bd2814f3cb65R669

> Was the customizable bounding box a requirement for this feature? Otherwise, we could start by always using the automatically sized bounding box and think about adding a padding size setting or a bounding box selector once it is requested thinking

This would have been alright for the first iteration probably, but when developing the feature I quickly noticed that a poor bounding box can lead to poor min cut results. Ideally, it should cover exactly the merger (not too much and not too little).  That's why I wanted to make it easy to select the bbox. If I would have known that this was going to be so tricky, I might have decided differently :sweat_smile: However, I'm confident that this context menu re-architecturing will be helpful in the future, too.